### ISSUE TITLE
Issues with BoxMatcher and loss

### ISSUE THREAD

**Issue by stiansel** [WongKinYiu/YOLO#103]
I saw the note in the readme about slower convergence and thought I'd try to help. These are the potential issues I've seen, though there may be others as well.

### BoxMatcher:

-  When doing normalization of align_cls the target_matrix hasn't been adjusted to account for the topk mask, and which GTs actually won the duplicate step. The same with the iou_mat. This yields different align_cls values in some assignments when comparing to other YOLO implementations.

-  Duplicate assignments in YOLO MIT are filtered based on the full cost matrix. Duplicate resolution in other single shot detector variants appear to use the (C)IoU cost only. Not sure how this affects the training.
 
-  While the no box case was fixed in #88 , there still might be a rare issue if there are target boxes, but they are all too small to overlap with the anchors. This may not be an issue unless doing custom datasets/loaders though.

### Loss

- The CIoU is missing no_grad around the penalty term. Adding it may improve loss stability. See https://github.com/pytorch/vision/blob/945bdad7523806b15d3740ce6ace2fced9ef9d3b/torchvision/ops/ciou_loss.py#L62

**Comment by henrytsui000** [comment]
Hi,

Thanks for raising this issue! I'm also working on figuring out why the convergence speed isn't as good as the stable version. Yesterday, I made some changes in commit fd5413f77d03f91b48eebba7dc1b98582bee93ad. If you have time, feel free to take a look.

Here’s a summary of the changes:
- Adjusted EPS from e-9 to e-7
- Switched to using the assignment matrix instead of the weighted matrix
- Modified the filtering logic to prioritize checking for duplicate bounding boxes

I believe some of these changes align with what you mentioned in this issue. Moving forward, I'll add no_grad to CIoU and implement logic to filter out extremely small bounding boxes. I’ll also retrain the model to check if the convergence issue still persists.

The changes so far only impacted the loss by about 0.001, so I suspect the issue may still exist. Let’s continue troubleshooting together!

Best regards,
Henry Tsui

**Comment by stiansel** [comment]
Sorry for the long delay. I've been away.

I think there may still be an issue with the duplicate function. Duplicates found tend to end up with low align_cls after normalization. I haven't looked at it closely yet, but I think the wrong indices may get picked?

```
# Code snippet from the BoxMatcher
topk_targets, topk_mask = self.filter_topk(target_matrix, topk=self.topk)

# If I check if any of the topk falls outside the grid mask:
print(torch.logical_and((grid_mask == 0), topk_mask).any())  # this line yields False

unique_indices, valid_mask, topk_mask = self.filter_duplicates(iou_mat, topk_mask)

# If the above had duplicates, and we check again, there appears to be occasional anchors that are assigned, but not in the grid mask (for that GT).
print(torch.logical_and((grid_mask == 0), topk_mask).any())  # this line yields True if we had duplicate assignments.
```

**Comment by Adamusen** [comment]
Hey @stiansel and @henrytsui000,

Thank you @stiansel for pointing out this location in the loss_function. With you finding out with the proposed check, that something goes wrong after the self.filter_duplicates() function call, I was able to further look in to it and eventually find out the mistake, which is rather trivial. The more complete code Snippet between Lines 266-281 inside yolo/utils/bounding_box_utils.py is as follows:

```
        # get valid matrix (each gt appear in which anchor grid)
        grid_mask = self.get_valid_matrix(target_bbox)

        # get iou matrix (iou with each gt bbox and each predict anchor)
        iou_mat = self.get_iou_matrix(predict_bbox, target_bbox)

        # get cls matrix (cls prob with each gt class and each predict class)
        cls_mat = self.get_cls_matrix(predict_cls.sigmoid(), target_cls)

        target_matrix = grid_mask * (iou_mat ** self.factor["iou"]) * (cls_mat ** self.factor["cls"])

        # choose topk
        topk_targets, topk_mask = self.filter_topk(target_matrix, topk=self.topk)

        # delete one anchor pred assign to mutliple gts
        unique_indices, valid_mask, topk_mask = self.filter_duplicates(iou_mat, topk_mask)
```

, which in essence first calculates the prerequisites for the "target_matrix" to be calculated, which is essencially a score, that tells how well each anchor is suited to each target in the ground truth (to predict it) based on class scores and IoU. The first use of said target_matrix is at "self.filter_topk", which makes sense, the following call, however, `self.filter_duplicates(iou_mat, topk_mask)` receives the "iou_mat" as it's first argument instead of the "target_matrix". This is intuitively a mistake, furher supported by the function definition of "self.filter_duplicates" at Lines 215-229:

```
    def filter_duplicates(self, target_matrix: Tensor, topk_mask: Tensor):
        """
        Filter the maximum suitability target index of each anchor.

        Args:
            target_matrix [batch x targets x anchors]: The suitability for each targets-anchors
            
        ...
```

, this mistake explains @stiansel observations of: 
1. "Duplicates found tend to end up with low align_cls after normalization."
2. "If the above had duplicates, and we check again, there appears to be occasional anchors that are assigned, but not in the grid mask (for that GT)", as the iou_mat is not multiplied by the "grid_mask".

As a consequence, sometimes anchors might got assignet to target boxes, which they are mathematically unable to predict (the anchor must be inside the target box to be able to predict it), and with incorrect target scores. This had obviously negative effects on the training, but I'm still not sure if this is/was THE Bug or just one of the bugs responsible for slow convergence :). In any case, I tested the fixed version during training and it seems to be better already, but further testing is still required to know if everything is good now or there still might be some bugs hidden causing slow convergence.

Pull Request incoming soon.

**Comment by Adamusen** [comment]
As a short follow up to my last comment: for a second I got confused which should be the better input for the filter_duplicates function the "target_matrix" or the "topk_targets", the latter which is never actually used. I tested it out with the following assertion:

```
        # delete one anchor pred assign to mutliple gts
        unique_indices, valid_mask, topk_mask = self.filter_duplicates(topk_targets, topk_mask)
        unique_indices2, valid_mask2, topk_mask2 = self.filter_duplicates(target_matrix, topk_mask)
        diff1 = torch.sum(unique_indices != unique_indices2)
        diff2 = torch.sum(valid_mask != valid_mask2)
        diff3 = torch.sum(topk_mask != topk_mask2)
        assert diff1 + diff2 + diff3 == 0, "It makes a difference!"
```

the assertion never triggered an error, so it doesn't matter.

**Comment by henrytsui000** [comment]
Hi @Adamusen,  

Using `iou_mat` as the input for `self.filter_duplicates` is intentional, but I did forget to update the function documentation—thanks for pointing that out!  

As @stiansel mentioned:  
> * Duplicate assignments in YOLO MIT are filtered based on the full cost matrix. Duplicate resolution in other single-shot detector variants appears to use the (C)IoU cost only. Not sure how this affects the training.

In light of this, I’ve updated the calculation matrix to use only IoU instead of the IoU & Class weight sum.

Best regards,  
Henry Tsui

**Comment by Adamusen** [comment]
Hi @henrytsui000,

thank you for your response, with this knowledge I know then, that my "quick fix" of switching back the input to the target_matrix on its own is not sufficient, however, the current implementation of simply using the iou_mat is still buggy. As @stiansel pointed out, the following check:

```
# If the above had duplicates, and we check again, there appears to be occasional anchors that are assigned, but not in the grid mask (for that GT).
print(torch.logical_and((grid_mask == 0), topk_mask).any())  # this line yields True if we had duplicate assignments.
```

sometimes produces True values (I tested it myself as well), which means, that sometimes anchor boxes are assigned to targets, which they are technically unable to predict, because the anchor is not inside the target box (signaled by grid_mask == 0). This problem stems most likely from the fact, that the "iou_mat" compared to the "target_matrix" is not yet masked by the "grid_mask", so even if one filters the duplicates based on the "iou_mat", it should be multiplied with the "grid_mask" beforehand to avoid this problem.
On top of that, if the duplicate filtering happens based on IoU only, the topk selection beforehand should be based on the same as well for the same of consistency.

In any case, I'm going to try to work myself in a little deeper in the theory of what is happening in this part of the loss_function in general based of other yolo implementations as well in the upcoming days and try to implement a proper clean fix then :)

**Comment by henrytsui000** [comment]
I'm in traffic, so quick reply, I tried doing topk_mask &= grid_mask after the last step of filter_duplicate, and training on my machine, it at least ensures the target within the anchor. Not sure whether the converge problem can be solved.

**Comment by Adamusen** [comment]
I'm going to be experimenting with this as well in the coming days :)

But as a quick note: I would multiple the iou_mat with the grid mask before calling the filter_duplicate function instead of afterwards, because if you do it after, you basically remove the wrong target from the anchor, which leaves it without any target, even tho it be able assigned a proper target (some other one from the duplicates, which it is suited for), but I might get something wrong here, I need to improve my understading of what exactly is happening here first :)

**Comment by Adamusen** [comment]
Hey @henrytsui000,

I checked the box matcher today in the "stable" version, and found out, that the bug, in which after the duplicate filtering an anchor might get assignet to a target, which it is not able to predict is contained in that implementation as well :D
Irrespective of that it is still a minor bug, which can easily be fixed by changing Line 281 in yolo/utils/bounding_box_utils.py to:

`unique_indices, valid_mask, topk_mask = self.filter_duplicates(grid_mask * iou_mat, topk_mask)`

which means: simply masking out all the iou values between gt_boxes and anchors, that can not be matched, because the anchor is not inside the gt_box.

Since with this knowledge this minor bug is basically certainly isn't the sole cause of slow convergence, I'll continue to walk through and control the rest of the loss_function on Monday :)

**Comment by henrytsui000** [comment]
Hi @Adamusen  

I believe the main issue might not be in the loss function or box matcher. I'm training my model, loss function, and post-process using the stable version framework, and the convergence speed is slightly better than the stable version so far (up to the 75th epoch). This leads me to suspect that the issue could be related to one of the following:  
- Validation metrics  
- Learning rate schedule  
- Optimization strategy  
- Gradient clipping  
- Loss scaling  

Additionally, I noticed that the convergence state in our framework is not only lower but also unstable—current performance can occasionally be worse than the previous epoch, not sure what is the reason. (By the way, using `topk_mask &= grid_mask` seems to make it a bit more stable.)

**Comment by Adamusen** [comment]
Hi @henrytsui000

Today I finished going through the entire loss_function and BoxMatcher in detail, comparing it with the stable version as well, and at this point found no more differences in the two. However, I found a relatively complex issue, which is present in both implementations, but I still find it a problem and will open a new, separate question issue about it soon.

In addition, I looked in to the items which you listed in your last reply and here is my summary:
 - **Validation metrics:** Since the lightning update, the repo uses the official implementation of  torchmetrics.detection.MeanAveragePrecision, so I don't think there should be any issues there. For me, the validation metrics look properly calculated. The only problem I could miss, if there is an issue with the original validation data loader, because I use my custom loader.
 - **Learning rate schedule:** There is an actually quite big issue here! This might be even the answer to the convergence problems. I will open a new, separate ticket about it soon and will be testing the training with the fix starting today.
 - **Optimization strategy:** One could try a different optimizer, but SGD is the default in the stable version as well, so it should work in the MIT version just as well.
 - **Gradient clipping:** It is slightly different to the stable version currently, as the lightning module uses "norm" instead of "value" as its default gradient_clip_algorithm. It could be updated to be identical to the stable version by using:
 `gradient_clip_algorithm="value"` as an extra initialization parameter for the Trainer in lazy.py, as described in the official lightning page for [gradient clipping](https://lightning.ai/docs/pytorch/stable/advanced/training_tricks.html#gradient-clipping).
 - **Loss scaling:** I noticed that you have recently updated the loss_values to be scaled with the batch size. This is part of the stable version as well, so at this point loss scaling should be identical in both versions.

About the unstable convergence: Yes, I have noticed the same issue.

**Comment by henrytsui000** [comment]
Hi @Adamusen,

I tried disabling EMA for the stable version and found that the model's performance (without EMA) is similar to the current repo's performance.

I have to admit that I'm not very familiar with EMA operations, and the existing Lightning + EMA implementations are under the Apache License, not MIT. I'm also uncertain if my implementation might cause a memory leak. However, I'll try the following implementation tomorrow:

```python
class EMA(Callback):
    def __init__(self, decay: float = 0.9999):
        super().__init__()
        logger.info(":chart_with_upwards_trend: Enable Model EMA")
        self.decay = decay

    def setup(self, trainer, pl_module, stage):
        self.original_state_dict = pl_module.state_dict()
        self.ema_state_dict = deepcopy(self.original_state_dict)

    @rank_zero_only
    def on_train_batch_end(self, trainer: "Trainer", pl_module: LightningModule, *args, **kwargs) -> None:
        with torch.no_grad():
            for key, value in pl_module.state_dict().items():
                ema_value = self.ema_state_dict[key]
                ema_value.copy_((1.0 - self.decay) * value + self.decay * ema_value, non_blocking=True)

    def on_train_epoch_start(self, trainer, pl_module):
        pl_module.load_state_dict(self.original_state_dict)

    def on_train_epoch_end(self, trainer, pl_module):
        self.original_state_dict = pl_module.state_dict()
        pl_module.load_state_dict(self.ema_state_dict)
```
I’ll let you know how it works out!

Best regards,
Henry Tsui

**Comment by Adamusen** [comment]
Hey @henrytsui000,

I personally don't think EMA is worth it overall, it makes the training significantly slower and in the end, the EMA and "normal" weight values should be converging to the same. In theory it only helps to achieve better validation scores during early in training, but should be the same after convergence eventually. I am still intrigued by the results, so please keep me updated :)

Two small notes to your implementation (looks good tho, should work without memory leaks):
 - The line `self.original_state_dict = pl_module.state_dict()` in "on_train_epoch_end" is basically redundant, it is still the same reference :)
 - You can do the update without needing to do a copy with an assign_subtract as described in the tensorflow documentation [here](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage), by reformulating the equation to:
`ema_value -= (1 - self.decay) * (ema_value - value)`

As a follow up: What do you think about issue #122?

**Comment by Adamusen** [comment]
Hey @henrytsui000 ,

as an update to my last reply about EMA: On the way home I realized, that your solution won't work properly, because you have to make a deepcopy from the original state_dict values on_train_epoch_end before you load the ema values, otherwise the original values will be lost forever, as your self.original_state_dict is nothing but a reference to the same state_dict values in which the ema values are then being loaded. In addition, to avoid gradient and graph connection problems, one should likely make a copy of each tensor in the state_dict one by one and do a .detach() on each copied tensor similar to how EMA was implemented previously.
